### ISSUE TITLE
Double indentation of class body when superclass is wrapped

### ISSUE THREAD

**Issue by TWiStErRob** [ktlint/ktlint#2423]
## Expected Behavior
Input is valid. Class body is correctly indented. Final brace is aligned with class opener.
Maybe the `SuperClass(primaryParam) {` line indented, but not the whole body. IntelliJ also just does this one line indent.

## Observed Behavior
```kotlin
internal abstract class MyClass internal constructor(primaryParam: ResourceFolderRepository) :
    SuperClass(primaryParam) {
        private val privateProperty: File = primaryProperty.resourceDir

        fun doStuff() {
            val file = privateProperty.resolve("foo")
        }
    }
```

## Steps to Reproduce
`ktlint "MyClass.kt" --format`
```kotlin
internal abstract class MyClass internal constructor(primaryParam: ResourceFolderRepository) :
SuperClass(primaryParam) {
    private val privateProperty: File = primaryProperty.resourceDir

    fun doStuff() {
        val file = privateProperty.resolve("foo")
    }
}
```

Note: when I unwrap the first line:
```kotlin
internal abstract class MyClass internal constructor(primaryParam: ResourceFolderRepository) : SuperClass(primaryParam) {
```
the indentation does not change at all, neither with:
```kotlin
internal abstract class MyClass internal constructor(
    primaryParam: ResourceFolderRepository
) : SuperClass(primaryParam) {
```

But this header also reproduces this issue:
```kotlin
internal abstract class MyClass internal constructor(primaryParam: ResourceFolderRepository)
: SuperClass(primaryParam) {
```

## Your Environment

* Version of ktlint used: 1.0.1
* Relevant parts of the `.editorconfig` settings: none
* Name and version (or code for custom task) of integration used (Gradle plugin, Maven plugin, command line, custom Gradle task): ktlint CLI
* Version of Gradle used (if applicable): N/A
* Operating System and version: Windows 10

**Comment by paul-dingemans** [comment]
This is intended behavior for code style `ktlint_official` which is the default code style since `1.0`.  See:
* https://ktlint.github.io/ktlint/latest/rules/code-styles/
* https://ktlint.github.io/ktlint/latest/faq/#can-a-new-toggle-be-added-to-optionally-enabledisable-format-code-in-a-particular-way

[edit: fixed links]

**Comment by TWiStErRob** [comment]
What is the rationale behind re-indenting hundreds of lines based only on a very few lines of class header? I guess that part of the answer is that it matches some level of logical indentation, but which part? (I'm really curious about this!)

---

Here's a reason for not doing this:

When I'm looking at any random code at line 230, and see 2 levels of indentation it gives implicit context that the code is nested inside something. This rule takes away this crutch and requires anyone reading any code formatted with ktlint_official to actually have to look at the top level definition (which might be at the top of the file, or might be at line 145) to figure out if this is the case. This requires a lot of mental and physical (scrolling) gymnastics for something that's a given in 99.9% of auto-formatted code styles in a plethora of languages not just Kotlin.

---

Note: I might be actually fine with this, because I always advocate for the clearly structured:
```
class Name(
    param: Type, // any number of times
) : SuperClass(...) {
```
style which gives a clean and immediate overview without the need for brain-parsing the code. And this style doesn't reproduce this issue. It just feels weird to reformat a whole file based on the placement of `:`
 * `) : Super` -> "normal" indent
 * `)\n: Super` -> "double" indent
 * `) :\nSuper` -> "double" indent

**Comment by diasDominik** [comment]
I also agree with @TWiStErRob the content belongs to the class and not the superclass so have the double indent makes it less clear were the code belongs to. Also I don't see a bigger class means that it is automatically a code smell(at which point is it actually a bigger class).

I'm all in for consistency but at the same time we should also not forget that the code should not get more complicated to read for humans.

Coming from python this indentiation would also be wrong so depends from where you come. But having it with a single indent is i believe in all languages the same

**Comment by itera-brodmo** [comment]
I have to agree with the OP, and even go a step further, I think this makes ktlint basically unusable for common scenarios. With test code like this
```kotlin
class Test(
    val mockService: Service
) : WordSpec({
    val test = 0
})
```
which you will often have when using Spring, the body is double indented with no way to disable that behavior other than disabling the standard indent rule. Having thousands of lines of test code double intended is just not an option, especially since KoTest already adds another level of indention with most specs.

**Comment by bric3** [comment]
Actually, I stumbled on a similar issue that has the primary constructor annotated. See #3193

> When I'm looking at any random code at line 230, and see 2 levels of indentation it gives implicit context that the code is nested inside something. This rule takes away this crutch and requires anyone reading any code formatted with ktlint_official to actually have to look at the top level definition (which might be at the top of the file, or might be at line 145) to figure out if this is the case.

👍 

Additionally

1. This causes IntelliJ (using the editor formater) and ktlint (used via gradle) fighting, which is very annoying.
2. CI is regularly failing for this reason.
3. The indentation is inconsistent with other classes.


**tl;dr:** workaround is to **NOT** use ktlint official rules, see https://github.com/ktlint/ktlint/issues/3193#issuecomment-3570502447

**Comment by AGDownie** [comment]
For what it's worth, I agree with the OP that this behaviour is erroneous, makes no logical sense, causes confusion and simply looks wrong.

I can't access the Pinterest links explaining the rationale behind this design choice (I get redirected to an "Engineering" landing page), but I would argue that the logic is flawed for the reasons the other contributors have outlined.  

Additionally, the indentation is not consistent.  The following two classes are semantically identical, but are formatted differently.  Why?
```
// correctly indented
class Test1(
    val s: String,
) : View.OnClickListener {
    override fun onClick(p0: View?) {
        println(s)
    }
}

// incorrectly indented
class Test2
    constructor(
        val s: String,
    ) : View.OnClickListener {
        override fun onClick(p0: View?) {
            println(s)
        }
    }
```
The first example correctly associates the braces and indentation level with the class.  The second example instead appears to erroneously associate the braces and indentation with the constructor clause.

Please un-park and fix this issue, so that we can start using ktlint in "Distract free" mode again.  It's quite the reverse at the moment as this distracts me pretty badly!  To the extent that I've disabled automatic ktlint in Android Studio and now occasionally use ktlint manually on an ad hoc basis to do a general tidy up, then fall back to IntelliJ formatting to fix this error.

This shouldn't detract from the great work that's gone into developing ktlint and the fact that it is an extremely useful tool. So thanks for all your efforts.

**Comment by paul-dingemans** [comment]
> I can't access the Pinterest links explaining the rationale behind this design choice (I get redirected to an "Engineering" landing page), but I would argue that the logic is flawed for the reasons the other contributors have outlined.

I have fixed the links:
* https://ktlint.github.io/ktlint/latest/rules/code-styles/
* https://ktlint.github.io/ktlint/latest/faq/#can-a-new-toggle-be-added-to-optionally-enabledisable-format-code-in-a-particular-way

> Please un-park and fix this issue

Sorry, you bring no new arguments to the table that convince me to fix the issue.

> Please un-park and fix this issue, so that we can start using ktlint in "Distract free" mode again. It's quite the reverse at the moment as this distracts me pretty badly! To the extent that I've disabled automatic ktlint in Android Studio and now occasionally use ktlint manually on an ad hoc basis to do a general tidy up, then fall back to IntelliJ formatting to fix this error.

If the indentation problem is holding you back from using the Ktlint Intellij Plugin, you might consider to disable the indentation rule (at your own risk of course): https://ktlint.github.io/ktlint/latest/faq/#how-do-i-enable-or-disable-a-rule. 

> This shouldn't detract from the great work that's gone into developing ktlint and the fact that it is an extremely useful tool. So thanks for all your efforts.

Thank you that you still value Ktlint for the good stuff that it brings as well ;-)

**Comment by AGDownie** [comment]
Thanks for the quick response and fixing the links.  But I don't think those provide a rationale for this particular issue.

In terms of not bringing any new arguments, you haven't answered my question why the two examples I provided are formatted differently. They are semantically identical, so the formatting can't be correct in both cases.  I would maintain the second should follow the same pattern as the first, i.e. single indent.

**Comment by paul-dingemans** [comment]
> Thanks for the quick response and fixing the links. But I don't think those provide a rationale for this particular issue.

As written in https://github.com/ktlint/ktlint/issues/2423#issuecomment-1850341689:
> > What is the rationale behind re-indenting hundreds of lines based only on a very few lines of class header? I guess that part of the answer is that it matches some level of logical indentation, but which part? (I'm really curious about this!)
> 
> The idea is to write _all_ class headers in a consistent way. There are a few pain points when doing so:
> 
> * annotations can be placed at many different elements in the class header. Not only on top of the class, but also on an explicit constructor, on type parameters, on class parameters, etc. From a consistency viewpoint, I want to handle annotions the same way, regardless of the construct on which they are used.
> * different indenting rules were applies on type parameters (with an indented `>`), versus class parameters (with de-indented `)`)
> 
> Is it worth to indent the entire class body compared to the size of the class header? That depends on your context and the size of your classes. For small classes it doesn't hurt at all. For big classes (code smell) it might actually hurt.

> In terms of not bringing any new arguments, you haven't answered my question why the two examples I provided are formatted differently. They are semantically identical, so the formatting can't be correct in both cases. I would maintain the second should follow the same pattern as the first, i.e. single indent.

From Ktlint perspective it is not relevant that your two classes are semantically identical. The identation of the class body is bound to the indentation of the `constructor` keyword. So when the `constructor` keyword is on the same line as the `class` keyword like below, it is accepted by ktlint without reformatting:
```
class Test2 constructor(
    val s: String,
) : View.OnClickListener {
    override fun onClick(p0: View?) {
        println(s)
    }
}
```

But if you wrap the `constructor` keyword to a separate line, it means that this keywords needs to be indented relative to the `class` keyword which in turn affects the indentation of the class body.

**Comment by AGDownie** [comment]
> From Ktlint perspective it is not relevant
> that your two classes are semantically
> identical. 

It should be relevant, though. In both cases the body contained in the braces pertains to the class object, not the constructor declaration, so there should be no extra indentation. 

Just because one uses the primary constructor pattern instead of an explicit constructor shouldn't alter the indentation of the class body. It should also be irrelevant whether the constructor is wrapped or not: the constructor has no body, so there is nothing to indent.

But I think we'll probably just have to agree to disagree on this. 

It doesn't cause any major difficulties; just a minor blemish on an otherwise extremely useful tool.  

Thanks again for all your good work.

**Comment by bcmedeiros** [comment]
@paul-dingemans the argument here is very elemental, do you disagree that the constructor and the class body are siblings in the AST? Both are below the class, which is the parent.
If you agree with that, then it's a natural consequence of that that siblings should not affect each other.

if you have a class with let's say 2000 lies, how do you justify to yourself introducing an indentation to all 2000 lines because you added an annotation to the constructor?

One of the main advantages of ktlint in comparison to ktfmt is how respectful and coherent it is of the current formatting of the class, but this one rule is the complete opposite of that, very intrusive and incoherent.

**Comment by paul-dingemans** [comment]
> @paul-dingemans the argument here is very elemental, do you disagree that the constructor and the class body are siblings in the AST? Both are below the class, which is the parent.

Indeed they are siblings.

> If you agree with that, then it's a natural consequence of that that siblings should not affect each other.

I do not agreee with this reasoning.

> if you have a class with let's say 2000 lies, how do you justify to yourself introducing an indentation to all 2000 lines because you added an annotation to the constructor?

The number of lines of the class is not relevant to this discussion.



I still do not agree with any arguments regarding the double indentation. But as I am really growing tired of the discussion, I will add a new property `ktlint_indent_explicit_constructor` that can be used to set the behavior regarding indentation of the `constructor` keyword. The default behavior for code style `ktlint_official` will be indenting the constructor keyword and class body like is done in Ktlint 1.x.
